README
NAME

    FFI::Platypus::Lang::Rust - Documentation and tools for using Platypus
    with the Rust programming language

SYNOPSIS

    Rust:

     #![crate_type = "dylib"]
     
     // compile with: rustc add.rs
     
     #[no_mangle]
     pub extern "C" fn add(a:i32, b:i32) -> i32 {
         a+b
     }

    Perl:

     use FFI::Platypus 1.00;
     my $ffi = FFI::Platypus->new( api => 1, lang => 'Rust' );
     $ffi->lib('./libadd.so');
     
     $ffi->attach( add => ['i32', 'i32'] => 'i32' );
     
     print add(1,2), "\n";  # prints 3

DESCRIPTION

    This module provides native Rust types for FFI::Platypus in order to
    reduce cognitive load and concentrate on Rust and forget about C types.
    This document also documents issues and caveats that I have discovered
    in my attempts to work with Rust and FFI.

    This module is somewhat experimental. It is also available for adoption
    for anyone either sufficiently knowledgeable about Rust or eager enough
    to learn enough about Rust. If you are interested, please send me a
    pull request or two on the project's GitHub.

    Note that in addition to using pre-compiled Rust libraries, you can
    bundle Rust code with your Perl distribution using FFI::Build and
    FFI::Build::File::Cargo.

 name mangling

    Rust names are "mangled" to handle features such as modules and the
    fact that some characters in Rust names are illegal machine code symbol
    names. For now that means that you have to tell Rust not to mangle the
    names of functions that you are going to call from Perl. You can
    accomplish that like this:

     #[no_mangle]
     pub extern "C" fn foo() {
     }

    You do not need to add this decoration to functions that you do not
    directly call from Perl. For example:

     fn bar() {
     }
     
     #[no_mangle]
     pub extern "C" fn foo() {
         bar();
     }

 panics

    Be careful about code that might panic!. A panic! across an FFI
    boundary is undefined behavior. You will want to catch the panic with a
    catch_unwind and map to an appropriate result.

     use std::panic::catch_unwind;
     
     #[no_mangle]
     pub extern fn oopsie() -> u32 {
         let result = catch_unwind(|| {
             might_panic();
         });
         match result {
             OK(_) => 0,
             Err(_) -> 1,
         }
     }

 structs

    You can map a Rust struct to a Perl object by creating a C OO layer. I
    suggest using the c_void type aliased to an appropriate name so that
    the struct can remain private to the Rust code.

    For example, given a Foo struct:

     struct Foo {
         ...
     }
     
     impl Foo {
         fn new() -> Foo { ... }
         fn method1(&self) { ... }
     }

    You can write a thin C layer:

     type CFoo = c_void;
     
     #[no_mangle]
     pub extern "C" fn foo_new(_class *const i8) -> *mut CFoo {
         Box::into_raw(Box::new(Foo::new())) as *mut CFoo
     }
     
     #[no_mangle]
     pub extern "C" fn foo_method1(f: *mut CFoo) {
         let f = unsafe { &*(f as *mut Foo) };
         f.method1();
     }
     
     #[allow(non_snake_case)]
     #[no_mangle]
     pub extern "C" fn foo_DESTROY(f: *mut CFoo) {
         unsafe { drop(Box::from_raw(f as *mut Foo)) };
     }

    Which can be called easily from Perl:

     package Foo {
     
         use FFI::Platypus 1.00;
         my $ffi = FFI::Platypus->new( api => 1, lang => 'Rust' );
         $ffi->bundle; # see FFI::Build::File::Cargo for how to bundle
                       # your rust code...
         $ffi->type( 'object(Foo)' => 'CFoo' );
         $ffi->mangler(sub {
             my $symbol = shift;
             "foo_$symbol";
         });
         $ffi->attach( new     => [] => 'CFoo' );
         $ffi->attach( method1 => ['CFoo'] );
         $ffi->attach( DESTROY => ['CFoo'] );
     };
     
     my $foo = Foo->new;
     $foo->method1;
     # $foo->DESTROY implicitly called when it falls out of scope

 returning strings

    Passing in strings is not too hard, you can convert a Rust CString into
    a Rust String. Return a string is a little tricky because of the
    ownership model. Depending on how your API works there are probably lot
    of approaches you might want to take. One approach would be to use
    thread local storage to store a CString which you return. It wastes a
    little memory because once the string is copied into Perl space it
    isn't used again, but at least it doesn't leak memory since it will be
    freed on the next call to your function. Best of all it doesn't require
    an unsafe block.

     pub extern "C" fn return_string() -> *const i8 {
         thread_local! {
             static KEEP: RefCell<Option<CString>> = RefCell::new(None);
         }
     
         let my_string = String::from("foo");
         let c_string = CString::new(my_string).unwrap();
         let ptr = c_string.as_ptr();
         KEEP.with(|k| {
             *k.borrow_mut() = Some(c_string);
         });
     
         ptr;
     }

    From Perl:

     use FFI::Platypus 1.00;
     my $ffi = FFI::Platypus->new( api => 1, lang => 'Rust' );
     $ffi->bundle;
     $ffi->attach( return_string => [] => 'string' );

 callbacks

    Calling back into Perl from Rust is easy so long as you have the
    correct types defined. Consider a Rust function that takes a C function
    pointer:

     use std::ffi::CString;
     
     type PerlLog = extern fn(line: *const i8);
     
     #[no_mangle]
     pub extern "C" fn rust_log(logf: PerlLog) {
     
         let lines: [&str; 3] = [
             "Hello from rust!",
             "Something else.",
             "The last log line",
         ];
     
         for line in lines.iter() {
             // convert string slice to a C style NULL terminated string
             let line = CString::new(*line).unwrap();
             logf(line.as_ptr());
         }
     }

    This can be called with a closure from Perl:

     use FFI::Platypus 1.00;
     
     my $ffi = FFI::Platypus->new( api => 1, lang => 'Rust' );
     $ffi->bundle;
     $ffi->type( '(string)->void' => 'PerlLog' );
     $ffi->attach( rust_log => ['PerlLog'] );
     
     my $perl_log = $ffi->closure(sub {
         my $message = shift;
         print "log> $message\n";
     });
     
     rust_log($perl_log);

    Which outputs:

     $ perl callback.pl
     log> Hello from rust!
     log> Something else.
     log> The last log line

METHODS

    Generally you will not use this class directly, instead interacting
    with the FFI::Platypus instance. However, the public methods used by
    Platypus are documented here.

 native_type_map

     my $hashref = FFI::Platypus::Lang::Rust->native_type_map;

    This returns a hash reference containing the native aliases for the
    Rust programming languages. That is the keys are native Rust types and
    the values are libffi native types.

EXAMPLES

    See the above "SYNOPSIS" or the examples directory that came with this
    distribution. This distribution comes with a whole module example of a
    full object-oriented Rust/Perl extension including Makefile.PL Rust
    crate, Perl library and tests. It lives in the examples/Person
    directory, or you can browse it on the web here:

    https://github.com/PerlFFI/FFI-Platypus-Lang-Rust/tree/main/examples/Person

SUPPORT

    If something does not work as advertised, or the way that you think it
    should, or if you have a feature request, please open an issue on this
    project's GitHub issue tracker:

    https://github.com/PerlFFI/FFI-Platypus-Lang-Rust/issues

CONTRIBUTING

    If you have implemented a new feature or fixed a bug then you may make
    a pull request on this project's GitHub repository:

    https://github.com/PerlFFI/FFI-Platypus-Lang-Rust/issues

    Caution: if you do this too frequently I may nominate you as the new
    maintainer. Extreme caution: if you like that sort of thing.

    This project's GitHub issue tracker listed above is not Write-Only. If
    you want to contribute then feel free to browse through the existing
    issues and see if there is something you feel you might be good at and
    take a whack at the problem. I frequently open issues myself that I
    hope will be accomplished by someone in the future but do not have time
    to immediately implement myself.

    Another good area to help out in is documentation. I try to make sure
    that there is good document coverage, that is there should be
    documentation describing all the public features and warnings about
    common pitfalls, but an outsider's or alternate view point on such
    things would be welcome; if you see something confusing or lacks
    sufficient detail I encourage documentation only pull requests to
    improve things.

SEE ALSO

    FFI::Platypus

      The Core Platypus documentation.

    FFI::Build::File::Cargo

      Bundle Rust code with your FFI / Perl extension.

AUTHOR

    Graham Ollis <plicease@cpan.org>

    Contributors:

    Andrew Grangaard (SPAZM)

COPYRIGHT AND LICENSE

    This software is copyright (c) 2015 by Graham Ollis.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.17      2023-08-22 13:44:06 -0600
  - Fixed documentation error that was listing Rust source
    twice instead of the Rust and Perl source as it should
    have done (gh#43, gh#44)

0.16      2023-06-14 17:05:55 -0600
  - use `*std::os::raw::c_char` instead of `*i8` for strings,
    as this is more portable.  In particular *u8 or *c_char
    is expected on Linux Arm64. (gh#42)

0.15      2023-02-21 08:05:34 -0700
  - Bump Alien::Rust prereq to 0.03 for better handling
    of rust home.

0.14      2023-02-16 06:16:37 -0700
  - Use Alien::Rust if rustc and cargo are not already
    installed (gh#39, gh#40)

0.12      2022-10-27 21:54:54 -0600
  - Documentation improvements (gh#38)

0.11      2022-10-26 05:57:14 -0600
  - Fix bug where FFI::Build::File::Cargo would usually exit
    on failure instead of die, which in dev / prove mode would
    throw away any diagnostics (gh#37)

0.10      2022-10-23 19:39:14 -0600
  - Documentation improvements (gh#5, gh#17, gh#22, gh#28,
    gh#31, gh#32, gh#35)
  - Adding bool type.  As noted in the CAVEATS, this is in
    practice always a sint8, but not technically guaranteed
    by the spec (gh#25, gh#33)
  - Use cdylib in examples and in implementation as that is
    preferred for calling from C now (gh#16, gh#31)
  - Tests are build with --release by default, just like the
    runtime cdylib, so that the Rust code doesn't have to be
    compiled twice (gh#26, gh#30)
  - The PERL_FFI_CARGO_FLAGS environment variable can be used
    to change the flags passed into cargo test and cargo build
    this way you can use a different profile if you choose
    (gh#30)

0.09 2019-11-20 11:12:37 -0700
  - Document callbacks.
  - Documentation fixes.

0.08 2019-11-19 16:02:53 -0700
  - Fix memory leak in points example (gh#5, gh#10)
  - Added bundle example: examples/Person
  - Better and more complete documentation for both
    FFI::Build::File::Cargo and
    FFI::Platypus::Lang::Rust.

0.07 2019-11-15 13:59:52 -0700
  - Migrate to ExtUtils::MakeMaker
  - Migrate from Module::Build::FFI to FFI::Build
  - Added FFI::Build::File::Cargo

0.06 2019-06-19 18:34:38 -0400
  - Documentation updates and fixes

0.05 2018-12-19 14:40:25 -0700
  - Explicitly make File::chdir a configure and runtime prereq

0.04 2015-10-14
  - Fixes for Mac OS X

0.03 2015-10-13
  - If there is a Cargo.toml file in the ffi project directory, use
    cargo to build and test the project.
  - The old method of populating ffi and libtest with rust files
    (not widely used) will be deprecated and removed at the end
    of the year.

0.02 2015-02-02
  - Fix synopsis to use "lang" instead of "with"

0.01 2015-01-30
  - Initial release


Makefile.PL
use strict;
use warnings;

BEGIN {
  unless(eval q{ use 5.008001; 1 }) {
    print "Perl 5.008001 or better required\n";
    exit;
  }
}

use 5.008001;
use lib 'lib';
use ExtUtils::MakeMaker 6.64;
use File::Which qw( which );
use FFI::Build::MM;

unless(which('cargo'))
{
  print "This distribution requires that you have cargo/rust installed\n";
  exit;
}

my $fbmm = FFI::Build::MM->new;

my %WriteMakefileArgs = (
  NAME               => 'FFI::Platypus::Lang::Rust',
  DISTNAME           => 'FFI-Platypus-Lang-Rust',
  AUTHOR             => "Graham Ollis <plicease\@cpan.org>",
  VERSION_FROM       => 'lib/FFI/Platypus/Lang/Rust.pm',
  ABSTRACT_FROM      => 'lib/FFI/Platypus/Lang/Rust.pm',
  LICENSE            => 'perl',
  MIN_PERL_VERSION   => '5.008001',
  CONFIGURE_REQUIRES => {
    'ExtUtils::MakeMaker' => '6.64',
    'File::Which'         => 0,
    'FFI::Build::MM'      => '1.00',
    'FFI::CheckLib'       => '0.11',
    'File::chdir'         => 0,
    'Path::Tiny'          => 0,
  },
  BUILD_REQUIRES => {
    # ...
  },
  PREREQ_PM => {
    'File::chdir'         => 0,
    'Path::Tiny'          => 0,
    'FFI::Platypus'       => '1.00',
    'FFI::CheckLib'       => '0.11',
  },
  TEST_REQUIRES => {
    'Capture::Tiny'       => 0,
    'Test::Script'        => 0,
  },
  META_MERGE => {
    'meta-spec' => { version => 2 },
    resources => {
      bugtracker => {
        web => 'https://github.com/PerlFFI/FFI-Platypus-Lang-Rust/issues',
      },
      repository => {
        type => 'git',
        url => 'https://github.com/PerlFFI/FFI-Platypus-Lang-Rust.git',
        web => 'https://github.com/PerlFFI/FFI-Platypus-Lang-Rust',
      },
      x_IRC => 'irc://irc.perl.org/#native',
    },
    no_index => {
      directory => [ 'examples' ],
    },
  },
  clean => { FILES => 'FFI-Platypus-Lang-Rust-* t/ffi/target examples/*.so examples/*.dll examples/*.dylib examples/Person/ffi/_build examples/Person/ffi/target' },
);

WriteMakefile($fbmm->mm_args(%WriteMakefileArgs));

sub MY::postamble {
  $fbmm->mm_postamble . "\n\nreadme:lib/FFI/Platypus/Lang/Rust.pm\n\tperl inc/run/readme.pl\n\n";
}
